### PR TITLE
Use openvino to implement openpose

### PR DESCRIPTION
**Pull request by playboy007gy**
Implement the models that openpose supported, including body pose, face, and hand.
You can easily change the model type to be used by annotating the code in "human_pose_estimator.cpp", "peak.cpp", "render_human_pose.cpp"; and recompile, like

#define COCO
//#define MPI
//#define BODY_25
//#define FACE
//#define HAND

The gitbub of openpose is: https://github.com/CMU-Perceptual-Computing-Lab/openpose
You can download openpose models followed the link.
It is worth noting that have to remove the last concat layer before MO if the model is coco or mpi.

### CODE REVIEW

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp [hunk 1]
```diff
@@ -7,14 +7,40 @@
...
 #include <samples/common.hpp>
 
 #include "human_pose_estimator.hpp"
 #include "peak.hpp"
 
+#define COCO
+//#define MPI
+//#define BODY_25
+//#define FACE
+//#define HAND
```
Comment (dkurt) [inline demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: It's better to introduce enum instead compile time definitions. So the same code can work with different types of networks (i.e. both body pose and hand).

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp [hunk 2]
```diff
@@ -42,12 +68,23 @@ HumanPoseEstimator::HumanPoseEstimator(const std::string& modelPath,
     network = netReader.getNetwork();
     InferenceEngine::InputInfo::Ptr inputInfo = network.getInputsInfo().begin()->second;
     inputLayerSize = cv::Size(inputInfo->getTensorDesc().getDims()[3], inputInfo->getTensorDesc().getDims()[2]);
-    inputInfo->setPrecision(InferenceEngine::Precision::U8);
```
Comment (dkurt) [inline demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: Please try to keep U8 input due it's more efficient.

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp [hunk 3]
```diff
@@ -89,44 +130,102 @@ std::vector<HumanPose> HumanPoseEstimator::estimate(const cv::Mat& image) {
...
+	//	<< pafsBlob->getTensorDesc().getDims()[0] << ", " 
+	//	<< pafsBlob->getTensorDesc().getDims()[1] << ", " 
+	//	<< pafsBlob->getTensorDesc().getDims()[2] << ", "
+	//	<< pafsBlob->getTensorDesc().getDims()[3] << std::endl;
+	//std::cout << "heatmapsBlobName = " << heatmapsBlobName << "   " 
+	//	<< heatMapsBlob->getTensorDesc().getDims()[0] << ", "
+	//	<< heatMapsBlob->getTensorDesc().getDims()[1] << ", "
+	//	<< heatMapsBlob->getTensorDesc().getDims()[2] << ", "
+	//	<< heatMapsBlob->getTensorDesc().getDims()[3] << std::endl;
+
```
Comment (dkurt) [inline demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: Please remove commented code.

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp (hunk 1 above)
Comment (playboy007gy) [reply demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: I rewrote the code, now can use
./humen_estimation_demo -i [cam/image] -m [mode] -t [model_type]
to run, which mode_type including COCO/MPI/BODY_25/FACE/HAND.

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp (hunk 2 above)
Comment (playboy007gy) [reply demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: Done

File: demos/human_pose_estimation_demo/src/human_pose_estimator.cpp (hunk 3 above)
Comment (playboy007gy) [reply demos/human_pose_estimation_demo/src/human_pose_estimator.cpp]: Done

File: demos/human_pose_estimation_demo/include/human_pose_estimation_demo.hpp [hunk 4]
```diff
@@ -13,8 +13,11 @@ static const char help_message[] = "Print a usage message.";
 /// @brief Message for video argument
 static const char video_message[] = "Required. Path to a video. Default value is \"cam\" to work with camera.";
 
-/// @brief Message for model argument
-static const char human_pose_estimation_model_message[] = "Required. Path to the Human Pose Estimation model (.xml) file.";
+/// @brief Message for openpose model argument
+static const char openpose_model_message[] = "Required. Path to the Human Pose Estimation model (.xml) file.";
```
Comment (dkurt) [inline demos/human_pose_estimation_demo/include/human_pose_estimation_demo.hpp]: Let's name it as `pose_estimation_model_message`

Comment (playboy007gy) [reply demos/human_pose_estimation_demo/include/human_pose_estimation_demo.hpp]: OK, I had finished the modification.

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Comment by Daniil-Osokin** [comment]
@playboy007gy Hand and face models do not work in this way. This is multi-person pose estimation, have you tried to run face or hand model on images with multiple persons (take any from MS COCO dataset), does hand/face model works well? In OpenPose they detect full pose at first, then make crops of hands and faces, based on detected skeleton keypoints locations, resizes them and pases to face/hand networks. So, I believe current results for face/hand will be quite unsatisfactory to show.

Also why do you need these models, e.g. MPI model is the same as from MS COCO, but with fewer number of keypoints, why it is useful to support it?

**Comment by playboy007gy** [comment]
> @playboy007gy Hand and face models do not work in this way. This is multi-person pose estimation, have you tried to run face or hand model on images with multiple persons (take any from MS COCO dataset), does hand/face model works well? In OpenPose they detect full pose at first, then make crops of hands and faces, based on detected skeleton keypoints locations, resizes them and pases to face/hand networks. So, I believe current results for face/hand will be quite unsatisfactory to show.
> 
> Also why do you need these models, e.g. MPI model is the same as from MS COCO, but with fewer number of keypoints, why it is useful to support it?

Hand and face models are only allowed for the images with only one object (face/hand).
If you give the required image, the model will work. Of cause we can also combined multi-models to get face/hand ROI by pose estimation, and then get face/hand features on their models. I can modify a version of multi-model if needed.

COCO/MPI/BODY_25 provide users a choice, below I have listed the performance of the supported models before and after using OpenVINO.
MPI model has fewer feature points and higher performance. (but it's slower than COCO in OpenVINO...)
BODY_25 model has the largest number of feature points, but it also has the lowest performance.
Overall, the performance of all of these models will be improved by the  deployment on OpenVINO.

--------------------------------------------------------------------
                             OPENPOSE (fps)                OpenVINO (fps)
COCO                            0.28                                   2.12
MPI                                0.38                                   1.25
BODY_25                       0.14                                    0.61
FACE                              0.37                                    1.2
HAND                            0.43                                   0.92
---------------------------------------------------------------------
Devicce |
CPU - Intel® Core™ i5-8350U CPU

**Comment by Daniil-Osokin** [comment]
Thanks for the numbers! You see, performance is uncool, looks like it's not very useful.

**Comment by playboy007gy** [comment]
Well, because the model provided by openpose has not been optimized, it's conceivable that the performance is not good enough.
My first contribution here is to extend the demo to  face and hand.
In addition, if users have the relevant technical capabilities, they  can also refer to the existing model to redesign the model architecture with less computation.
For example, the model provided by OpenVINO "human-pose-estimation-0001" is similar to COCO.

**Comment by Daniil-Osokin** [comment]
So, how this demo will be different from the one, provided in OpenPose? I would love to see proper face/hand integration, but with reasonable performance and accuracy.

**Comment by playboy007gy** [comment]
> So, how this demo will be different from the one, provided in OpenPose? I would love to see proper face/hand integration, but with reasonable performance and accuracy.

Sorry to make you confused, I didn't clearly explain the devices used in the performance test.
Since OpenPose can only be used CUDA (Nvidia GPU), OpenCL (AMD GPU), and CPU-only (no GPU); and OpenVINO can run on Intel CPU, Intel Processor Graphics, MyriadX products.
So the data comparison mentioned above is OpenPose in GPU (GeForce GTX 1080 Ti) and OpenVINO in CPU (CPU - Intel® Core™ i5-8350U CPU).

I have done one other test recently, which uses "video.avi" as the sample and "body_25" as the test model. However, OpenPose runs on ”Intel(R) Core(TM) i9-7900X CPU” and ”GeForce GTX 1080 Ti” respectively, and OpenVINO only runs on ”Intel(R) Core(TM) i9-7900X CPU”.
Screenshots of the test results and the sample videos have been uploaded to this patch.

                          OPENPOSE-CPU (fps)               OPENPOSE-GPU (fps)
 ------------------------------- 0.047---------------------------------- 4.7

                          OpenVINO-CPU (fps)
 ------------------------------- 3.78

If you run OpenPose in CPU-only, there will be a serious delay between the frames.
Due to openVINO's model optimizer can convert the model into a low-precision mode, it can make the fps in using CPU close to in using GPU.

In addition, I am doing integrated facial/hand programming, which will be released upon completion.

**Comment by Daniil-Osokin** [comment]
Thanks for the tests. Open Model Zoo is for accurate and fast models, so this PR may be stuck until one provide such models.

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

**Comment by vladimir-dudnik** [comment]
@playboy007gy seems you need to rebase to the latest state of develop branch.
[ ERROR ] Parameter -t is not set

**Comment by playboy007gy** [comment]
> @playboy007gy seems you need to rebase to the latest state of develop branch.
> [ ERROR ] Parameter -t is not set

[-t] is a new option to set model type. You can run it followed
### ./humen_estimation_demo -i [cam/image] -m [mode] -t [model_type]
, which mode_type including COCO/MPI/BODY_25/FACE/HAND.

**Comment by IRDonch** [comment]
Alright, so:

1. The discussion appears to have stalled here; and
2. I have looked at the linked repository, and the content there is licensed under a non-open source license. This means that we can't integrate their models into OMZ.

As such, I don't think this is going to get anywhere. Closing.

**Pull request closed without merging**